### PR TITLE
[spoiler] feat: adding Spoiler plugin for image, gallery and video

### PR DESCRIPTION
**Pull request by barboaron**

### CODE REVIEW

File: packages/editor-common/web/src/Base/createBaseComponent.jsx [hunk 1]
```diff
@@ -187,6 +188,12 @@ const createBaseComponent = ({
       }
     };
 
+    onComponentSpoilerChange = data => {
+      if (this.isMeAndIdle()) {
+        this.updateComponentConfig({ spoiler: data });
```
Comment (shaulgo) [inline packages/editor-common/web/src/Base/createBaseComponent.jsx:193]: spoiler: {
enabled: true/false,
description: ...
}

File: packages/editor-common/web/src/Base/toolbars/createAtomicPluginToolbar.jsx [hunk 2]
```diff
@@ -116,6 +123,11 @@ export default function createAtomicPluginToolbar({
       pubsub.update('componentData', { config: { link } });
     };
 
+    onComponentSpoilerChange = data => {
+      const spoiler = data;
+      pubsub.update('componentData', { config: { spoiler } });
```
Comment (shaulgo) [inline packages/editor-common/web/src/Base/toolbars/createAtomicPluginToolbar.jsx:128]: why is this needed both here and in createBaseComponent.jsx?

File: packages/editor-common/web/src/Base/toolbars/createAtomicPluginToolbar.jsx [hunk 3]
```diff
@@ -223,6 +235,19 @@ export default function createAtomicPluginToolbar({
         toolbarOffsetLeft: this.state.position && this.state.position['--offset-left'],
         innerModal,
       };
+
+      const baseSpoilerProps = {
+        tabIndex,
```
Comment (shaulgo) [inline packages/editor-common/web/src/Base/toolbars/createAtomicPluginToolbar.jsx:240]: let's make a commonButtonProps

File: packages/editor-common/web/src/Base/toolbars/createAtomicPluginToolbar.jsx [hunk 4]
```diff
@@ -282,6 +307,15 @@ export default function createAtomicPluginToolbar({
           return <Separator className={separatorClassNames} horizontal key={index} />;
         case BUTTONS.LINK:
           return <BlockLinkButton {...baseLinkProps} tooltipText={t('TextLinkButton_Tooltip')} />;
+        case BUTTONS.SPOILER:
+          return (
+            <BlockSpoilerButton
+              {...baseSpoilerProps}
+              getEditorState={getEditorState}
+              setEditorState={setEditorState}
```
Comment (shaulgo) [inline packages/editor-common/web/src/Base/toolbars/createAtomicPluginToolbar.jsx:315]: check if needed

File: packages/plugin-gallery/web/src/gallery-viewer.jsx [hunk 5]
```diff
@@ -228,27 +229,33 @@ class GalleryViewer extends React.Component {
     const viewMode = this.props.seoMode ? GALLERY_CONSTS.viewMode.SEO : undefined;
 
     return (
-      <div
-        ref={elem => (this.container = elem)}
-        className={this.styles.gallery_container}
-        data-hook={'galleryViewer'}
-        role="none"
-        onContextMenu={this.handleContextMenu}
+      <BlockSpoilerComponent
```
Comment (shaulgo) [inline packages/plugin-gallery/web/src/gallery-viewer.jsx:232]: try to use decorators instead of having the spoiler in other plugins.

File: CHANGELOG.md [hunk 6]
```diff
@@ -27,7 +27,8 @@
 
 ### :rocket: New Feature
 - `spoiler`
-  - [#1194](https://github.com/wix-incubator/rich-content/pull/1194) Spoiler plugin for text
+  - [#1194](https://github.com/wix-incubator/rich-content/pull/1194) adding spoiler plugin for text
```
[the remark this answers is not in the document]
Comment (barboaron) [reply CHANGELOG.md:30]: done

File: e2e/cypress/support/commands.js [hunk 7]
```diff
@@ -339,6 +339,8 @@ Cypress.Commands.add('openGalleryAdvancedSettings', () => {
   cy.get(`[data-hook=${PLUGIN_COMPONENT.GALLERY}]:first`)
     .parent()
     .click();
+  cy.get('[data-hook="pluginToolbarRightArrow"]:first').click();
+  cy.wait(200);
   cy.get(`[data-hook=${PLUGIN_TOOLBAR_BUTTONS.ADV_SETTINGS}]:first`).click();
```
[the remark this answers is not in the document]
Comment (barboaron) [reply e2e/cypress/support/commands.js:344]: done

File: e2e/tests/plugins.e2e.js [hunk 8]
```diff
@@ -56,27 +56,80 @@ describe('plugins', () => {
     });
 
     after(() => cy.eyesClose());
-    it(`check spoilers in editor`, () => {
+    it(`check spoilers on text in editor`, () => {
```
[the remark this answers is not in the document]
Comment (barboaron) [reply e2e/tests/plugins.e2e.js:59]: done

File: e2e/tests/plugins.e2e.js [hunk 9]
```diff
@@ -56,27 +56,80 @@ describe('plugins', () => {
...
-      cy.setTextStyle('spoilerButton', [30, 10]);
+      cy.setTextStyle('textSpoilerButton', [30, 10]);
       cy.eyesCheckWindow('adding some spoilers');
       cy.setLink([5, 5], 'https://www.wix.com/');
-      cy.setTextStyle('spoilerButton', [0, 13]);
+      cy.setTextStyle('textSpoilerButton', [0, 13]);
       cy.eyesCheckWindow('adding spoiler around link');
-      cy.setTextStyle('spoilerButton', [20, 10]);
+      cy.setTextStyle('textSpoilerButton', [20, 10]);
       cy.eyesCheckWindow('union spoilers');
```
[the remark this answers is not in the document]
Comment (barboaron) [reply e2e/tests/plugins.e2e.js:71]: done

File: e2e/tests/plugins.e2e.js [hunk 10]
```diff
@@ -56,27 +56,80 @@ describe('plugins', () => {
     });
 
     after(() => cy.eyesClose());
-    it(`check spoilers in editor`, () => {
+    it(`check spoilers on text in editor`, () => {
       cy.loadRicosEditorAndViewer('empty', usePlugins(plugins.spoilerPreset)).enterParagraphs([
         'Leverage agile frameworks to provide a robust synopsis for high level overviews. Iterative approaches to corporate strategy foster collaborative thinking to further the overall value proposition.',
       ]);
-      cy.setTextStyle('spoilerButton', [15, 5]);
+      cy.setTextStyle('textSpoilerButton', [15, 5]);
```
[the remark this answers is not in the document]
Comment (barboaron) [reply e2e/tests/plugins.e2e.js:63]: done

File: e2e/tests/plugins.e2e.js [hunk 11]
```diff
@@ -56,27 +56,80 @@ describe('plugins', () => {
...
+      cy.setTextStyle('textSpoilerButton', [20, 10]);
       cy.eyesCheckWindow('union spoilers');
-      cy.setTextStyle('spoilerButton', [20, 5]);
+      cy.setTextStyle('textSpoilerButton', [20, 5]);
       cy.eyesCheckWindow('split spoiler');
     });
 
-    it(`reveal spoiler in viewer`, () => {
+    it(`reveal spoiler on text in viewer`, () => {
       cy.get('[data-hook="spoiler_0"]:first').click();
```
[the remark this answers is not in the document]
Comment (barboaron) [reply e2e/tests/plugins.e2e.js:77]: done

File: e2e/tests/plugins.e2e.js [hunk 12]
```diff
@@ -56,27 +56,80 @@ describe('plugins', () => {
...
-      cy.setTextStyle('spoilerButton', [20, 10]);
+      cy.setTextStyle('textSpoilerButton', [20, 10]);
       cy.eyesCheckWindow('union spoilers');
-      cy.setTextStyle('spoilerButton', [20, 5]);
+      cy.setTextStyle('textSpoilerButton', [20, 5]);
       cy.eyesCheckWindow('split spoiler');
     });
 
-    it(`reveal spoiler in viewer`, () => {
+    it(`reveal spoiler on text in viewer`, () => {
```
[the remark this answers is not in the document]
Comment (barboaron) [reply e2e/tests/plugins.e2e.js:76]: done

File: e2e/tests/plugins.e2e.js [hunk 13]
```diff
@@ -56,27 +56,80 @@ describe('plugins', () => {
...
       cy.get('[data-hook="spoiler_0"]:first').click();
       cy.eyesCheckWindow('reveal spoiler');
     });
+
+    it(`check spoilers on an image in editor and reveal it in viewer`, () => {
+      cy.loadRicosEditorAndViewer('images');
+      cy.get('[data-hook="imageViewer"]:first')
+        .parent()
+        .click();
+      cy.get('[data-hook="spoilerButton"]:visible').click();
```
[the remark this answers is not in the document]
Comment (barboaron) [reply e2e/tests/plugins.e2e.js:86]: done

File: packages/editor-common/web/src/Base/createBaseComponent.jsx (hunk 1 above)
Comment (barboaron) [reply packages/editor-common/web/src/Base/createBaseComponent.jsx:193]: done

File: packages/editor-common/web/src/Base/toolbars/createAtomicPluginToolbar.jsx (hunk 2 above)
Comment (barboaron) [reply packages/editor-common/web/src/Base/toolbars/createAtomicPluginToolbar.jsx:128]: removed from createAtomicPluginToolbar

File: packages/editor-common/web/src/Base/toolbars/createAtomicPluginToolbar.jsx (hunk 3 above)
Comment (barboaron) [reply packages/editor-common/web/src/Base/toolbars/createAtomicPluginToolbar.jsx:240]: done

File: packages/editor-common/web/src/Base/toolbars/createAtomicPluginToolbar.jsx (hunk 4 above)
Comment (barboaron) [reply packages/editor-common/web/src/Base/toolbars/createAtomicPluginToolbar.jsx:315]: deleted
